### ISSUE TITLE
V2 possible modules approach

### ISSUE THREAD

**Issue by mshgh** [jorgebucaran/hyperapp#773]
I have moved the idea I mentioned here https://github.com/jorgebucaran/hyperapp/issues/749#issuecomment-432956449 little further.
Here is live working demo https://stackblitz.com/edit/ha2-slices-demo

Goals I wanted to achieve
- couple fragments of _state_ with corresponding _actions_ and _effects_ (I call this **slice**)
- use **slice**s to define _state_ and make them reusable (e.g. to have two counters in a page)
- components responsible for _state_ visualization are pure (provided with _props_ and _props_ only)
- avoid need to propagate _state_ top-down (use React/Redux like **connect** approach)
  - **connected** component has no (or bare minimum) UI in it

In the demo
- all under **/npm/...** becomes npm package later
- all under **/lib/...** is reusable building blocks ready to be shared among projects
- and under **/app/...** is the application itself; _state_ definition and _connected_ components

**Comment by SkaterDad** [comment]
I've done a little digging into the code to understand what you're doing, but would be interested in more explanation since it's a lot to look at.

Could you explain how your `connect` component is receiving the state?  It looks like it comes from a function param, but in my limited time I can't trace where it gets injected.

I would love if `hyperapp` would continue to support that feature (`const Component = (props, children) => (state) => h('div', {}, 'Whatever')`) in V2, as it is in V1.

**Comment by mshgh** [comment]
Sure. In order to allow slice actions to work on the state sub-tree they are wrapped in `map()` function which is generated by the `squirrel()` function. This is the approach I took from discussion in the referenced issue.
```js
export default map => {
  const inc = amount => map(state => ({
      ...state, 
      count: state.count + 1,
  }));
  return { inc };
```
The trick in my `add()` method (responsible to register slice) is to not provide directly the `map()` generated by `squirrel()` but wrap it into another function which keeps reference of generated sub-state in `sliceInfo` object. The `sliceInfo` is then stored into global object called `slices` using the same `map()` method to ensure the same hierarchy for `sliceInfo`s and in the state. The whole idea is in the /npm/@hyperapp-v2/slices.js the rest is just to show it woks with actions and effects too.
Here is code fragment from the slices.js which shows what I tried to describe above.
```js
let slices = {}; // all added slices are stored here
const add = (path, slice) => { // path defines structure e.g. "conters.counter1" 
  const sliceInfo = {}; // object to keep info about slice
  // this generates method allowing to updated any object based on provided path
  const map = squirrel(path);
  // here is the trick - generated map() is wrapped in a way I get access to generated sub-state
  // and I keep it in the sliceInfo object.
  // slice.actions is the exported method in code fragment above.
  const { _init = s => s, ...actions } = slice.actions(action => map(state => sliceInfo.state = action(state)));
  // I want to make the "adjusted" actions available for the application
  sliceInfo.actions = actions; // it's bit different in the demo because there are supported effects too. this is simplification to get the idea
  // here is the sliceInfo stored into slices using the same map() method a.k.a. path
  slices = map(_ => sliceInfo)(slices);
}
```
Later the `connect()` method simply provides `slices` to the mapToProps function.
Let say you call `add("foo.bar",someSlice)` then the `mapToProps` receives `slices` where you can access state and actions of the `someSlice` via `slices.foo.bar.state` and `slices.foo.bar.actions`
I hope I am making sense... There is bit more in place to support effects.

In a nutshell I keep the state fragments in the `slices` object with the same hierarchy as the `state` has. I think this is not problem to read the state info from this paralel storage as this shares the same references as the `state` itself.

**Comment by SkaterDad** [comment]
> @mshgh In a nutshell I keep the state fragments in the slices object with the same hierarchy as the state has. I think this is not problem to read the state info from this paralel storage as this shares the same references as the state itself.

This is what I suspected, thanks for clarifying!

While I find this clever, I'm not sure I would end up using it in my apps due to the complexity of the implementation.

As discussed in another topic, functional lenses would be able to accomplish something similar to slices?  I use a **ton** of slices in the SPA I'm building, so will definitely be reaching for a migration path.

I've also been considering [Microstates](https://github.com/microstates/microstates.js/), which has a really neat approach to immutable data models.  Now that hyperapp v2 doesn't auto-merge an object, we can potentially use this library.

**Comment by zaceno** [comment]
@SkaterDad, @mshgh 

FWIW this HOA + function lets you connect components to state in HAV2, similar to how "lazy-components" worked in v1. It's arguably cleaner than storing a reference to the state. And it can easily be adapted to working with mapped state:

```js
    const resolve = (node, state) => {
        if (node == null) {
            //just return it
        } else if (Array.isArray(node)) {
            node = node.map(function (n) { return resolve(n, state)})
            node = Array.prototype.concat.apply([], node)
            node = node.filter(function (n) { return n != null })
        } else if (node.name === '$') {
            node = resolve(node.props.f(state), state)
        } else if (node.children) {
            node.children = resolve(node.children, state)
        } else {
            node = {name: '' + node, props: {}, children: [], type: 2, element: undefined, key: null}
        } 
        return node
    }

    export const view = f => ({name: '$', props: {f}, children: [], type: 0, element: undefined, key: null})

    export const withViews = app => def => app({
        ...def,
        view: state => resolve(view(def.view), state)
    })

```

Demo: https://codepen.io/zaceno/pen/VEOKpV?editors=0011

**Comment by zaceno** [comment]
@SkaterDad 
>As discussed in another topic, functional lenses would be able to accomplish something similar to slices? 

Yes, totally. The `squirrel` function is basically poor-mans-lensing. And yes, the idea is to solve slices somehow. I personally haven't used ramda or similar, and didn't want to pull in a new library just for this, so that's why I made squirrel. 

But for the purpose of porting an app over from v1 to v2, I don't think any of these solutions matter much. V2 is different enough it will likely require a lot rewriting and rearchitecting. You might want to look into @okwolf 's compatibility library.

**Comment by SkaterDad** [comment]
> @zaceno But for the purpose of porting an app over from v1 to v2, I don't think any of these solutions matter much. V2 is different enough it will likely require a lot rewriting and rearchitecting. You might want to look into @okwolf 's compatibility library.

I've been giving my app's architecture quite a lot of thought lately.  It's in need of a re-write, especially the data fetching stuff, but I'm so far in that it's a daunting task.  Reducing my reliance on slices will involve un-nesting a lot of my stuff, but it should be worth it in the end.

The V2 managed actions & effects make such nice looking code, that I'm looking forward to the migration.  Could possibly use `@hyperapp/fx` on my V1 app as a bridge to V2.

**Comment by lukejacksonn** [comment]
I have looked through the demo you posted.. what concerns me most about this approach is just how much code you have to write. To achieve what you want.. take this file for example:

```js
import { h } from '../../npm/hyperapp-v2'
import { connect } from '../../npm/@hyperapp-v2/slices'
import { Counter } from '../../lib/components'

const FruitCounters = ({
  applesCount, applesPending, applesInc, applesAdd, applesAddSome, applesAddLater, applesDec, applesSub,
  bananasCount, bananasPending, bananasInc, bananasAdd, bananasAddSome, bananasAddLater, bananasDec, bananasSub,
}) =>
  h('div', {}, [
    Counter({label: 'Apples',
      count:applesCount, pending: applesPending,
      inc:applesInc, add:applesAdd,
      addSome:[applesAddSome, {min:5, max:8}], addLater:[applesAddLater, {amount:2, after:2000}],
      dec:applesDec, sub:applesSub,
    }),
    Counter({label: 'Bananas',
      count:bananasCount, pending: bananasPending,
      inc:bananasInc, add:bananasAdd,
      addSome:[bananasAddSome, {min:9, max:11}], addLater:[bananasAddLater, {amount:3, after:1500}],
      dec:bananasDec, sub:bananasSub,
    }),
  ])

const mapToProps = slices => {
  const { applesCounting, bananasCounting } = slices;

  return {
    applesCount: applesCounting.state.count,
    applesPending: applesCounting.state.pending,
    applesInc: applesCounting.operations.increment,
    applesAdd: applesCounting.operations.add,
    applesAddSome: applesCounting.operations.addRandom,
    applesAddLater: applesCounting.operations.delayedAdd,
    applesDec: applesCounting.operations.decrement,
    applesSub: applesCounting.operations.sub,

    bananasCount: bananasCounting.state.count,
    bananasPending: bananasCounting.state.pending,
    bananasInc: bananasCounting.operations.increment,
    bananasAdd: bananasCounting.operations.add,
    bananasAddSome: bananasCounting.operations.addRandom,
    bananasAddLater: bananasCounting.operations.delayedAdd,
    bananasDec: bananasCounting.operations.decrement,
    bananasSub: bananasCounting.operations.sub,
  };
};

export default connect(mapToProps, FruitCounters);
```

There is _so_ much room for developer error here and so much naming of things with the actual names of things included across many files. Without types, I see this being pretty arduous to debug and refactor. I admit that "less code is better" is not always true..  but this solution is quite verbose. Have you thought about ways of reducing the verbosity here?

---

I have been playing with modules myself, trying to solve some of the same problems you mention in the issue description. I have ended up with something that allows you to define a "module" as some view wired to some namespace in state and some reducers you can run on that state.

A module, is pretty straightforward and looks something like this:

> modules/counter.js
```js
import { h } from 'https://rawgit.com/hyperapp/hyperapp/V2/src/index.js'

export const counter = (init = 0) => ({
  init,
  actions: {
    inc: state => state + 1,
    dec: state => state - 1,
  },
  view: state => actions =>
    h('div', {}, [
      h('h3', {}, `${props.label}: ${state}`),
      h('button', { onclick: actions.down }, 'Decrement'),
      h('button', { onclick: actions.up }, 'Increment'),
    ]),
})
```

Then assign this module to a namespace in state like this:

> index.js
```
import { counter } from './modules/counter.js'
import { modules } from './utils.js'

const { init, ...view } = modules({
  temperature: counter(27),
  humidity: counter(85),
})
```

This exposes two things.. the first is an `init` object which contains the initial state of all modules based on the namespace and values provided. In this case it would look like this:

```
{
  temperature: 27,
  humidity: 85,
}
```

The second thing exposed is a `view` object which contains the views wired to pluck out the relevant namespace in state and use the actions defined within the module. You can use them in your apps view like this:

```
view: state =>
    h('main', {}, [
      view.temperature(state),
      view.humidity(state),
    ]),
```

It seems to work quite well but certainly has some limitations.. for a start:

- Can't nest modules (I still haven't decided if this should be encouraged).
- Module actions are not exposed so only the wired views can reduce the namespace state.

I do wonder if anyone has tried any other way though. Or has thoughts about this approach as a less verbose alternative to above, can anyone see any glaring obvious flaws, is this feature even desirable? It is still very much a WIP so feedback is welcomed.

**Comment by mshgh** [comment]
@lukejacksonn the verbosity is outcome of few things
- this is PoC stage only
- it was created in small step via adding features by copy& paste, not with DRY in mind
- it is demo e.g. Counter is overloaded with features - like 6 buttons to update single number

My focus was on solving the listed "goals" and the source code was meant to help understand what I meant.
I checked on other ideas about modules before I wrote this. What have all the other proposal in common is strong binding of state, actions and view(s). I feel only state and action should be bound, and views should be used beyond this scope (or substate, or slice). However at the same time views should not understand the whole state either.

**Comment by SkaterDad** [comment]
> @mshgh . I feel only state and action should be bound, and views should be used beyond this scope (or substate, or slice).

I agree.  For me, it is rare that I have a totally isolated view component.  I often need to pass in other related state (and actions in V1).

**Comment by mshgh** [comment]
@lukejacksonn I challenged my approach with your sample and it turn out this was straightforward exercise. It appeared the state-action-views module to be just composition of what I already have at hand.
Here is identical (almost ;) module you provided
```js
const counterModule = {
  actions: {
    inc: () => state => state + 1,
    sub: amount => state => state - amount,
  },
  view: ({ label, state, api }) => {
    return h('div', {}, [
      h('button', { onClick: api.inc }, 'Inc'), ' ',
      h('button', { onClick: [api.sub, 5] }, 'Sub 5'), ' ',
      h('span', {}, `${label}: ${state}`),
    ]);
  }
}
```
Here is how you make views out of it
```js
const Temperature = module("temperature", counterModule, 27);
const Humidity = module("huminidity", counterModule, 85);
```
This way the views would be used
```js
app({
  init: ..., // omitted to prevent confusion
  view: state =>
    h('div', {}, [
      Temperature({ label: 'Temperature' }),
      Humidity({ label: 'Humidity' }),
    ]),
  container: document.getElementById('app'),
});
```
And here is implementation of the `module()` function (using already existing methods, actually)
```js
const module = (path, module, init) => {
  add(path, module, init);
  return connect(slices => ({ ...slices[path] }), module.view);
}
```
There would no problem to turn it into `modules()`-like syntax
```js
const views = modules({
  temperature: [counterModule, 27],
  humidity: [counterModule, 85],
});
```
More over there was no problem to extend the module with `effects`. No need for any additional changes beside this update to the module definition...
```js
const counterModule = {
  actions: {
    inc: () => state => state + 1,
    sub: amount => state => state - amount,
  },
  api: actions => ({
    ...actions,
    incLater: state => [
      state,
      Time({ action: actions.inc, after: 2000 })
    ],
  }),
  view: ({ label, state, api }) => {
    return h('div', {}, [
      h('button', { onClick: api.inc }, 'Inc'), ' ',
      h('button', { onClick: api.incLater }, 'Inc later'), ' ',
      h('button', { onClick: [api.sub, 5] }, 'Sub 5'), ' ',
      h('span', {}, `${label}: ${state}`),
    ]);
  }
}
```

My apology for this very long comment.
I am working on making my demo more robust and polished. Once I am done I prepare another live solution on _stackblitz_ so it is easier to review the implementation.

**Comment by lukejacksonn** [comment]
Ohh man, awesome work.. looking much cleaner already! 🎉

**Comment by mshgh** [comment]
Here is the new version at [stackblitz.com](https://stackblitz.com/edit/ha2-slices-demo-take2) If somebody wants to see the implementation of the modules itself have a look here [lib/hypeapp/modules.js](https://stackblitz.com/edit/ha2-slices-demo-take2?file=src%2Flib%2Fhyperapp%2Fmodules.js)

For those who might be interested how the usage in the application looks like I suggest to look at
- self contained (state, action, view) modul definition is here [lib/modules/control-panel.js](https://stackblitz.com/edit/ha2-slices-demo-take2?file=src%2Flib%2Fmodules%2Fcontrol-panel.js)
- modul with business logic only (state and actions) is here [lib/modules/counting.js](https://stackblitz.com/edit/ha2-slices-demo-take2?file=src%2Flib%2Fmodules%2Fcounting.js)
- sample of **view**( a.k.a. connected stateless component) can be seen at [app/views/pantry.js](https://stackblitz.com/edit/ha2-slices-demo-take2?file=src%2Fapp%2Fviews%2Fpantry.js) (note this is part of the application, not library)

The combination of all the modules with initial parameters is located in [app/app-setup.js](https://stackblitz.com/edit/ha2-slices-demo-take2?file=src%2Fapp%2Fapp-setup.js)

Opinions? ;)

**Comment by lukejacksonn** [comment]
Looks much better from a DRY perspective.. using tuples in `app-setup` is interesting.. can you see any potential shortcomings of allowing nesting like this `food.fruits`? I have ended up with something similar for generating lists of modules but something doesn't feel quite right about it :/ (not saying that it is wrong).

**Comment by mshgh** [comment]
I cannot see any problem coming with the `food.fruits` shortcut as this is nothing else but `['food', ['fruits', [ ... ]]]` replaced with `['food.fruits', [ ... ]]` as I know on the `food` level there is `fruits` only.

I like to have the module definition as tuple e.g. `['humidity', controlPanel, { desired: 85 }]` I think this is aligned with the **hav2** approach too.

Should I use tuples for namespaceing modules? I see few advantages
- minimal possible modules definition is `init: modules({foo: 'bar'}),` and it opens possibility to enhance the **ha2** itself to understand modules directly and the modules function can be removed e.g. `init: ['humidity', module, {baz: 'bar'}],` and it is backward compatible with `init:{foo: 'bar'},` It probably never happens, but still good this is possible ;)
- the modules can be interlaced with initial state definition. It is questionable if this is good idea to combine in single project slices approach with global state understanding approach. Still I didn't want to remove this possibility
- it allows for the namespace shortcuts `food.fruits` we discussed above

Changing the modules namespaceing to object structure, if module definition stays as tuple, is minor change to the overall concept, and if everybody feels this is better. Sure, why not.

P.S. I am thinking about simplification of effects definition in the module. I have some ideas, but I am afraid this is sort of bypassing **hav2**. I'll prepare updated demo to show more details.

**Comment by mshgh** [comment]
It was quite easy in the end. [New version](https://stackblitz.com/edit/ha2-slices-demo-take3) is ready. The updated module definitions are here [control-panel.js](https://stackblitz.com/edit/ha2-slices-demo-take3?file=src%2Flib%2Fmodules%2Fcontrol-panel.js) and [counting.js](https://stackblitz.com/edit/ha2-slices-demo-take3?file=src%2Flib%2Fmodules%2Fcounting.js). And my concern regarding impact to **hav2** was pointless.

**Comment by mshgh** [comment]
Oops! I realized there is cyclic dependency in the demo. Somehow this was not problem for build and running. It is fixed now by updating the [latest version](https://stackblitz.com/edit/ha2-slices-demo-take3). I had to export `views` instead to return them by modules() function. Just heads-up...
I consider it final version now. I do not have any ideas how to improve it further ;)

**Comment by mshgh** [comment]
I find out that my usage of effects in the demo was not the most straightforward one. What I am talking about. I turned _action_ into _effect_ and then used `BatchFx` to run sequence of effects (set pending to true; schedule delayed update). However there is more direct way in effects api - I can return new state right away. So I updated the modules implementation to cover this use case too. 

Here is how it end up.
Current approach. Still valid and useful, but only in cases state doesn't need to be updated before the effect. (in this particular case unnecessary complicated)
```js
addLater: ({ amount, after }) =>
  BatchFx(toEffect(actions.pending), Time({ action: [actions.add, amount], after })),
```
Newly added support doing the same job
```js
addLater: ({ amount, after }) =>
  [actions.pending,  Time({ action: [actions.add, amount], after })],
```
Or in case the action is taking any parameters...
```js
addLater: ({ amount, after }) =>
  [[actions.pending, true],  Time({ action: [actions.add, amount], after })],
```
I didn't update stackblizt.com with this small enhancement, but I can do it if anybody is interested.

**Comment by 132yse** [comment]
Oh, I think it's wonderful.
